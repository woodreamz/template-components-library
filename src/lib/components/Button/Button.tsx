import MuiButton, { ButtonProps as MuiButtonProps } from '@mui/material/Button';
import styled from 'styled-components';
import {
  getColor,
  getBackground,
  getColorHover,
  getBackgroundHover,
  getColorPress,
  getBackgroundPress,
  getBackgroundDisabled,
  getColorDisabled,
  getBorderColor,
  flavors,
  sizes,
  getSize,
  getIconSize,
  getFontSize,
  getBorderColorHover,
  getBorderColorPress,
  hexToRgba,
} from '../../theme/utils/styleUtils';
// @ts-nocheck

// #region constants
const defaultClasses = {
  root: 'buttonRoot',
  label: 'buttonLabel',
  focusVisible: 'buttonFocusVisible',
  disabled: 'buttonDisabled',
};
// #endregion

// #region functions
const getPadding = (size: any) => {
  let padding = '8px 16px';
  if (size === sizes.small) {
    padding = '4px 16px';
  }
  if (size === sizes.extraSmall) {
    padding = '2px 16px';
  }
  return padding;
};
// #endregion

// #region styled-components
const StyledButton = styled(MuiButton)<{
  $size: string;
  $flavor: string;
  variant: string;
  $svgStrokeWidth: number;
  $svgColorAttribute: string;
}>`
  &&.buttonRoot {
    min-height: ${({ $size }) => getSize($size)};
    height: ${({ $size }) => getSize($size)};
    font-size: ${({ $size }) => getFontSize($size)};
    line-height: 1.2em;
    text-transform: none;
    font-weight: ${({ theme }) => theme.fonts.semibold};
    padding: ${({ $size }) => getPadding($size)};
    color: ${({ $flavor, theme, variant }) => getColor(theme, variant, $flavor)};
    background-color: ${({ $flavor, theme, variant }) => getBackground(theme, variant, $flavor)};
    border: solid 1px ${({ $flavor, theme, variant }) => getBorderColor(theme, variant, $flavor)};
    box-shadow: none;
    /* No wrap and add ellipsis. */
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    display: inline-block;
    svg {
      height: ${({ $size, variant }) => getIconSize($size, variant)};
      width: ${({ $size, variant }) => getIconSize($size, variant)};
      fill: transparent;
      stroke: transparent;
      stroke-width: ${({ $svgStrokeWidth }) => $svgStrokeWidth};
      ${({ $svgColorAttribute }) => `${$svgColorAttribute}:currentColor`};
      flex-shrink: 0;
    }
  }
  &&.buttonRoot:hover {
    color: ${({ $flavor, theme, variant }) => getColorHover(theme, variant, $flavor)};
    background-color: ${({ $flavor, theme, variant }) =>
      getBackgroundHover(theme, variant, $flavor)};
    border: solid 1px
      ${({ $flavor, theme, variant }) => getBorderColorHover(theme, variant, $flavor)};
    svg {
      ${({ $svgColorAttribute }) => `${$svgColorAttribute}:currentColor`};
    }
  }
  &&.buttonRoot.buttonFocusVisible {
    box-shadow: 0px 0px 4px 0px
      ${({ theme }) => hexToRgba(theme.primaryFocus, theme.isLightTheme ? 0.5 : 1)};
  }
  &&.buttonRoot:active {
    color: ${({ $flavor, theme, variant }) => getColorPress(theme, variant, $flavor)};
    background-color: ${({ $flavor, theme, variant }) =>
      getBackgroundPress(theme, variant, $flavor)};
    border: solid 1px
      ${({ $flavor, theme, variant }) => getBorderColorPress(theme, variant, $flavor)};
    svg {
      ${({ $svgColorAttribute }) => `${$svgColorAttribute}:currentColor`};
    }
  }
  &&.buttonDisabled {
    color: ${({ theme, variant }) => getColorDisabled(theme, variant)};
    background-color: ${({ theme, variant }) => getBackgroundDisabled(theme, variant)};
    border: solid 1px transparent;
    svg {
      ${({ $svgColorAttribute }) => `${$svgColorAttribute}:currentColor`};
    }
  }
`;
// #endregion

// #region components
export interface ButtonProps extends MuiButtonProps {
  flavor?: string;
  svgColorAttribute?: string;
  svgStrokeWidth?: number;
}

const Button = ({
  children,
  classes,
  flavor = flavors.normal,
  svgColorAttribute = 'stroke',
  svgStrokeWidth = 1.5,
  ...others
}: ButtonProps) => {
  return (
    <StyledButton<any>
      {...others}
      $size={sizes.medium}
      classes={{ ...defaultClasses, ...classes }}
      $flavor={flavor}
      $svgColorAttribute={svgColorAttribute}
      $svgStrokeWidth={svgStrokeWidth}
      disableElevation
    >
      {children}
    </StyledButton>
  );
};
// #endregion

export default Button;
