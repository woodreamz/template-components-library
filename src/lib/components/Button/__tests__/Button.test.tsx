import { screen } from '@testing-library/react';
import { render } from '../../../../tests';
import Button from '../Button';

describe('<Button />', () => {
  it('should render correctly', () => {
    const { container } = render(<Button>Click here</Button>);
    expect(container).toMatchSnapshot();
  });

  it('should contain the good text', () => {
    const text = 'I am a button';
    render(<Button>{text}</Button>);
    screen.getByText(text);
  });

  it('should render text variant with normal flavor', () => {
    const { container } = render(
      <Button variant="text" flavor="normal">
        Click here
      </Button>
    );
    expect(container).toMatchSnapshot();
  });

  it('should render contained variant with normal flavor', () => {
    const { container } = render(
      <Button variant="contained" flavor="normal">
        Click here
      </Button>
    );
    expect(container).toMatchSnapshot();
  });

  it('should render outlined variant with normal flavor', () => {
    const { container } = render(
      <Button variant="outlined" flavor="normal">
        Click here
      </Button>
    );
    expect(container).toMatchSnapshot();
  });

  it('should render text variant with good flavor', () => {
    const { container } = render(
      <Button variant="text" flavor="good">
        Click here
      </Button>
    );
    expect(container).toMatchSnapshot();
  });

  it('should render contained variant with good flavor', () => {
    const { container } = render(
      <Button variant="contained" flavor="good">
        Click here
      </Button>
    );
    expect(container).toMatchSnapshot();
  });

  it('should render outlined variant with good flavor', () => {
    const { container } = render(
      <Button variant="outlined" flavor="good">
        Click here
      </Button>
    );
    expect(container).toMatchSnapshot();
  });

  it('should render text variant with attention flavor', () => {
    const { container } = render(
      <Button variant="text" flavor="attention">
        Click here
      </Button>
    );
    expect(container).toMatchSnapshot();
  });

  it('should render contained variant with attention flavor', () => {
    const { container } = render(
      <Button variant="contained" flavor="attention">
        Click here
      </Button>
    );
    expect(container).toMatchSnapshot();
  });

  it('should render outlined variant with attention flavor', () => {
    const { container } = render(
      <Button variant="outlined" flavor="attention">
        Click here
      </Button>
    );
    expect(container).toMatchSnapshot();
  });

  it('should render text variant with caution flavor', () => {
    const { container } = render(
      <Button variant="text" flavor="caution">
        Click here
      </Button>
    );
    expect(container).toMatchSnapshot();
  });

  it('should render contained variant with caution flavor', () => {
    const { container } = render(
      <Button variant="contained" flavor="caution">
        Click here
      </Button>
    );
    expect(container).toMatchSnapshot();
  });

  it('should render variant with caution flavor', () => {
    const { container } = render(
      <Button variant="outlined" flavor="caution">
        Click here
      </Button>
    );
    expect(container).toMatchSnapshot();
  });
});
