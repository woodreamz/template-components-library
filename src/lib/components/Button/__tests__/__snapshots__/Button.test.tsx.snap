// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Button /> > should render contained variant with attention flavor 1`] = `
<div>
  <button
    class="MuiButtonBase-root-JLQss eudYYC MuiButtonBase-root MuiButton-root buttonRoot MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-disableElevation MuiButton-root-gxrVZJ dFJbXi MuiButton-root buttonRoot MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-disableElevation sc-dlniIP jLnkCj"
    tabindex="0"
    type="button"
  >
    Click here
  </button>
</div>
`;

exports[`<Button /> > should render contained variant with caution flavor 1`] = `
<div>
  <button
    class="MuiButtonBase-root-JLQss eudYYC MuiButtonBase-root MuiButton-root buttonRoot MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-disableElevation MuiButton-root-gxrVZJ dFJbXi MuiButton-root buttonRoot MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-disableElevation sc-dlniIP gZIjSF"
    tabindex="0"
    type="button"
  >
    Click here
  </button>
</div>
`;

exports[`<Button /> > should render contained variant with good flavor 1`] = `
<div>
  <button
    class="MuiButtonBase-root-JLQss eudYYC MuiButtonBase-root MuiButton-root buttonRoot MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-disableElevation MuiButton-root-gxrVZJ dFJbXi MuiButton-root buttonRoot MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-disableElevation sc-dlniIP coPgxu"
    tabindex="0"
    type="button"
  >
    Click here
  </button>
</div>
`;

exports[`<Button /> > should render contained variant with normal flavor 1`] = `
<div>
  <button
    class="MuiButtonBase-root-JLQss eudYYC MuiButtonBase-root MuiButton-root buttonRoot MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-disableElevation MuiButton-root-gxrVZJ dFJbXi MuiButton-root buttonRoot MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-disableElevation sc-dlniIP jlWsQX"
    tabindex="0"
    type="button"
  >
    Click here
  </button>
</div>
`;

exports[`<Button /> > should render correctly 1`] = `
<div>
  <button
    class="MuiButtonBase-root-JLQss eudYYC MuiButtonBase-root MuiButton-root buttonRoot MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-disableElevation MuiButton-root-gxrVZJ hxiUDP MuiButton-root buttonRoot MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-disableElevation sc-dlniIP fHiWKo"
    tabindex="0"
    type="button"
  >
    Click here
  </button>
</div>
`;

exports[`<Button /> > should render outlined variant with attention flavor 1`] = `
<div>
  <button
    class="MuiButtonBase-root-JLQss eudYYC MuiButtonBase-root MuiButton-root buttonRoot MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-disableElevation MuiButton-root-gxrVZJ gdXmSA MuiButton-root buttonRoot MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-disableElevation sc-dlniIP bIUBKn"
    tabindex="0"
    type="button"
  >
    Click here
  </button>
</div>
`;

exports[`<Button /> > should render outlined variant with good flavor 1`] = `
<div>
  <button
    class="MuiButtonBase-root-JLQss eudYYC MuiButtonBase-root MuiButton-root buttonRoot MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-disableElevation MuiButton-root-gxrVZJ gdXmSA MuiButton-root buttonRoot MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-disableElevation sc-dlniIP gtNozz"
    tabindex="0"
    type="button"
  >
    Click here
  </button>
</div>
`;

exports[`<Button /> > should render outlined variant with normal flavor 1`] = `
<div>
  <button
    class="MuiButtonBase-root-JLQss eudYYC MuiButtonBase-root MuiButton-root buttonRoot MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-disableElevation MuiButton-root-gxrVZJ gdXmSA MuiButton-root buttonRoot MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-disableElevation sc-dlniIP gMExNX"
    tabindex="0"
    type="button"
  >
    Click here
  </button>
</div>
`;

exports[`<Button /> > should render text variant with attention flavor 1`] = `
<div>
  <button
    class="MuiButtonBase-root-JLQss eudYYC MuiButtonBase-root MuiButton-root buttonRoot MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-disableElevation MuiButton-root-gxrVZJ hxiUDP MuiButton-root buttonRoot MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-disableElevation sc-dlniIP kyWFfQ"
    tabindex="0"
    type="button"
  >
    Click here
  </button>
</div>
`;

exports[`<Button /> > should render text variant with caution flavor 1`] = `
<div>
  <button
    class="MuiButtonBase-root-JLQss eudYYC MuiButtonBase-root MuiButton-root buttonRoot MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-disableElevation MuiButton-root-gxrVZJ hxiUDP MuiButton-root buttonRoot MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-disableElevation sc-dlniIP ixmAXe"
    tabindex="0"
    type="button"
  >
    Click here
  </button>
</div>
`;

exports[`<Button /> > should render text variant with good flavor 1`] = `
<div>
  <button
    class="MuiButtonBase-root-JLQss eudYYC MuiButtonBase-root MuiButton-root buttonRoot MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-disableElevation MuiButton-root-gxrVZJ hxiUDP MuiButton-root buttonRoot MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-disableElevation sc-dlniIP gwkGwp"
    tabindex="0"
    type="button"
  >
    Click here
  </button>
</div>
`;

exports[`<Button /> > should render text variant with normal flavor 1`] = `
<div>
  <button
    class="MuiButtonBase-root-JLQss eudYYC MuiButtonBase-root MuiButton-root buttonRoot MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-disableElevation MuiButton-root-gxrVZJ hxiUDP MuiButton-root buttonRoot MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-disableElevation sc-dlniIP fHiWKo"
    tabindex="0"
    type="button"
  >
    Click here
  </button>
</div>
`;

exports[`<Button /> > should render variant with caution flavor 1`] = `
<div>
  <button
    class="MuiButtonBase-root-JLQss eudYYC MuiButtonBase-root MuiButton-root buttonRoot MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-disableElevation MuiButton-root-gxrVZJ gdXmSA MuiButton-root buttonRoot MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-disableElevation sc-dlniIP kJSHoL"
    tabindex="0"
    type="button"
  >
    Click here
  </button>
</div>
`;
